dev-docs/RFCs/proposals/transport-format-rfc.md

```diff
@@ -0,0 +1,110 @@
+# RFC: Transport Format RFC
+
+* Authors: Xiaoji Chen
+* Date: Jan, 2020
+* Status: **Draft**
+
+## Summary
+
+This RFC proposes a generic transport format for generating/rendering data cross platforms.
+
+## Background
+
+In v7.3, deck.gl added the `@deck.gl/json` module. It supports generating serializable deck.gl props in non-JavaScript environments and send it to render on the client side. This format is currently used by `pydeck`, deck.gl's Python binding.
+
+We have plans to add bindings for other languages, and potentially non-JS renderers. This calls for a more comprehensive format that can be easily generated, sent, and handled by multiple languages and platforms.
+
+The JSON format has the following shortcomings:
+
+- Efficiency: Serializing/deserializing large data objects with JSON is expensive in terms of both time and space. Network bandwidth, CPU time and memory are all valuable assets when it comes to web applications.
+- Updates: It is not possible to update a layer without re-sending the entire snapshot. This is important for animation and/or live data visualization.
+
+A hybrid transport format of both JSON and binary will help address these issues.
+
+## Proposal
+
+Expand the `JSONConverter.convert` API to accept a second argument:
+
+```js
+converter.convert(json, buffers);
+```
+
+Where `buffers` is a list/map of ArrayBuffers.
```

Comment:
Sending individual buffers is absolutely the reasonable starting point, though from an assumed typical user persona's perspective this could be considered an advanced use case. 

For binary data, it could be desirable to work towards supporting binary formats that contain all buffers and some or all of metadata required to access them) in a single file, i.e. Arrow (and possibly glTF), so that they could just be sent as a unit and then just supplied directly as `Layer.data`.

The canonical example being a pydeck user just needs to use `pyarrow` to convert his `pandas` data frame to a binary Arrow blob and then pass that blob as data to a deck.gl layer. Ideally the user has to provide as little mapping as possible on the deck.gl side. (The in-progress Arrow support RFCs have the beginnings of the required data mappings).

Reply:
Per my understanding of the current frameworks architecture, we don't want deck.gl to parse the Arrow format. Deck should understand the table category from loaders.gl. It's up to the application what format they want to use, and they can pick the corresponding loader, which outputs to the same, normalized format.

If that is the case, then there is no built-in support for "a single binary file." Deck will always receive a JSON descriptor, which contains accessors and arraybuffers. It is not deck's concern how a single binary is parsed to that JSON descriptor.